### PR TITLE
[C++ Worker]Python call cpp actor

### PR DESCRIPTION
**Pull request by qicosmos**
## Why are these changes needed?

[Last PR](https://github.com/ray-project/ray/pull/22820) has supported python call c++ normal task, this PR supports python call c++ actor task.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests

### CODE REVIEW

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

File: cpp/test_python_call_cpp.py [hunk 1]
```diff
@@ -60,6 +64,30 @@ def test_cross_language_cpp():
...
 
 
+def test_cross_language_cpp_actor():
+    actor = ray.cross_language.cpp_actor_class("CreateCounter", "Counter").remote()
+    obj = actor.Plus1.remote()
+    assert 1 == ray.get(obj)
+
+    actor1 = ray.cross_language.cpp_actor_class(
+        "RAY_FUNC(Counter::FactoryCreate)", "Counter"
+    ).remote()
```
Comment (ericl) [inline cpp/test_python_call_cpp.py:74]: Can you add tests for creating invalid actors?

Comment (qicosmos) [reply cpp/test_python_call_cpp.py:82]: Sure

**Comment by qicosmos** [comment]
The failed cases seem unrelated.

File: src/ray/common/function_descriptor.h [hunk 2]
```diff
@@ -230,23 +230,26 @@ class CppFunctionDescriptor : public FunctionDescriptorInterface {
 
   virtual size_t Hash() const {
     return std::hash<int>()(ray::FunctionDescriptorType::kCppFunctionDescriptor) ^
-           std::hash<std::string>()(typed_message_->function_name());
+           std::hash<std::string>()(typed_message_->function_name()) ^
```
Comment (mwtian) [inline src/ray/common/function_descriptor.h:233]: totally optional: I think absl and boost has hash combine implementations where order is taken into account.

**Review by mwtian — APPROVED** [review]

File: src/ray/common/function_descriptor.h (hunk 2 above)
Comment (qicosmos) [reply src/ray/common/function_descriptor.h:233]: Do you mean don't modify the Hash function?

File: cpp/src/ray/test/cluster/counter.h [hunk 3]
```diff
@@ -50,6 +50,9 @@ class Counter {
   bool is_restared = false;
 };
 
+static inline Counter *CreateCounter() { return new Counter(0); }
```
[the remark this answers is not in the document]
Comment (qicosmos) [reply cpp/src/ray/test/cluster/counter.h:53]: The static can be removed.

[the remark this answers is not in the document]
Comment (mwtian) [reply cpp/src/ray/test/cluster/counter.h]: I think in a separate PR, maybe some of the unnecessary `static` on function definitions can be removed.

File: src/ray/common/function_descriptor.h (hunk 2 above)
Comment (mwtian) [reply src/ray/common/function_descriptor.h:233]: It is ok and nothing needs to be changed in this PR. But I think using `^` to combine hashes makes (function_name="a", class_name="b") and (function_name="b", class_name="a") to produce the same output, which may not be desirable. absl and boost have utilities so hash(function_name="a", class_name="b") != hash(function_name="b", class_name="a").

File: cpp/src/ray/test/cluster/counter.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (qicosmos) [reply cpp/src/ray/test/cluster/counter.h]: Right.

**Comment by qicosmos** [comment]
Could anyone merge this PR?

**Pull request merged**